components/speedreader/tts_player.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/speedreader/tts_player.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "content/public/browser/tts_controller.h"
+#include "content/public/browser/tts_utterance.h"
+#include "content/public/browser/web_contents.h"
+
+namespace speedreader {
+
+TtsPlayer::TtsPlayer() = default;
+
+TtsPlayer::~TtsPlayer() = default;
+
+// static
+TtsPlayer* TtsPlayer::GetInstance() {
+  return base::Singleton<TtsPlayer>::get();
+}
+
+TtsPlayer::Controller& TtsPlayer::GetControllerFor(
+    content::WebContents* web_contents) {
+  controller_.SetRequestWebContents(web_contents);
+  return controller_;
+}
+
+void TtsPlayer::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void TtsPlayer::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+double TtsPlayer::GetSpeed() const {
+  return controller_.current_speed_;
+}
+
+const std::string& TtsPlayer::GetVoice() const {
+  return controller_.current_voice_;
+}
+
+void TtsPlayer::SetSpeed(double speed) {
+  if (std::abs(speed - controller_.current_speed_) > 0.05) {
+    controller_.current_speed_ = speed;
+    if (controller_.IsPlaying()) {
+      controller_.Resume(true);
+    }
+  }
+}
+
+void TtsPlayer::SetVoice(const std::string& voice) {
+  if (controller_.current_voice_ != voice) {
+    controller_.current_voice_ = voice;
+    if (controller_.IsPlaying()) {
+      controller_.Resume(true);
+    }
+  }
+}
+
+TtsPlayer::Controller::Controller(TtsPlayer* owner) : owner_(owner) {}
+TtsPlayer::Controller::~Controller() = default;
+
+void TtsPlayer::Controller::SetRequestWebContents(
+    content::WebContents* web_contents) {
+  request_web_contents_ = web_contents;
+}
+
+bool TtsPlayer::Controller::IsPlaying() const {
+  auto* tts = content::TtsController::GetInstance();
+  return tts->IsSpeaking();
+}
+
+bool TtsPlayer::Controller::IsPlayingRequestedWebContents() const {
+  return playing_web_contents_ == request_web_contents_;
+}
+
+void TtsPlayer::Controller::Play() {
+  DCHECK(request_web_contents_);
+  if (IsPlayingRequestedWebContents()) {
+    Observe(playing_web_contents_);
+    Resume(true);
+  } else {
+    Stop();
+    TtsPlayer::GetInstance()->delegate_->RequestReadingContent(
```

Comment:
Is this TTS player the same for all the websites?
Can it be deallocated?

Reply:
Yes it's browser-wide component. It can't be allocated.